lnrpc/invoicesrpc/addinvoice.go

```diff
@@ -678,169 +618,229 @@ type SelectHopHintsCfg struct {
-		if edgePolicy == nil || !canBeHopHint {
-			continue
-		}
+	return hopHints, nil
+}
 
-		// Similarly, in this first pass, we'll ignore all channels in
-		// isolation can't satisfy this payment.
-		if channel.RemoteBalance < amtMSat {
-			continue
-		}
+// getPotencialHints returns a slice of open channels that should be considered
+// for the hopHint list in an invoice. The slice is sorted in descending order
+// based on the remote balance.
+func getPotencialHints(cfg *SelectHopHintsCfg) ([]*channeldb.OpenChannel,
+	error) {
 
-		// Lookup and see if there is an alias SCID that exists.
-		chanID := lnwire.NewChanIDFromOutPoint(
-			&channel.FundingOutpoint,
-		)
-		alias, _ := cfg.GetAlias(chanID)
-
-		// If this is a channel where the option-scid-alias feature bit
-		// was negotiated and the alias is not yet assigned, we cannot
-		// issue an invoice. Doing so might expose the confirmed SCID
-		// of a private channel.
-		if channel.ScidAliasFeature {
-			var defaultScid lnwire.ShortChannelID
-			if alias == defaultScid {
-				continue
-			}
+	// TODO(positiveblue): get the channels slice already filtered by
+	// private == true and sorted by RemoteBalance?
+	openChannels, err := cfg.FetchAllChannels()
+	if err != nil {
+		return nil, err
+	}
+
+	privateChannels := make([]*channeldb.OpenChannel, 0, len(openChannels))
+	for _, oc := range openChannels {
+		isPublic := oc.ChannelFlags&lnwire.FFAnnounceChannel != 0
+		if !isPublic {
+			privateChannels = append(privateChannels, oc)
 		}
+	}
+
+	// Sort the channels in descending remote balance.
+	compareRemoteBalance := func(i, j int) bool {
+		iBalance := privateChannels[i].LocalCommitment.RemoteBalance
+		jBalance := privateChannels[j].LocalCommitment.RemoteBalance
+		return iBalance > jBalance
+	}
+	sort.Slice(privateChannels, compareRemoteBalance)
+
+	return privateChannels, nil
+}
+
+// shouldIncludeChannel returns true if the channel passes all the checks to
+// be a hopHint in a given invoice.
+func shouldIncludeChannel(channel *channeldb.OpenChannel,
+	alreadyIncluded map[uint64]struct{},
+	cfg *SelectHopHintsCfg) (zpay32.HopHint, lnwire.MilliSatoshi, bool) {
+
+	if _, ok := alreadyIncluded[channel.ShortChannelID.ToUint64()]; ok {
+		return zpay32.HopHint{}, 0, false
+	}
 
-		// Now that we now this channel use usable, add it as a hop
-		// hint and the indexes we'll use later.
-		addHopHint(&hopHints, channel, edgePolicy, alias)
+	chanID := lnwire.NewChanIDFromOutPoint(
+		&channel.FundingOutpoint,
+	)
+
+	hopHintInfo := newHopHintInfo(channel, cfg.IsChannelActive(chanID))
+
+	// If this channel can't be a hop hint, then skip it.
+	edgePolicy, canBeHopHint := chanCanBeHopHint(hopHintInfo, cfg)
+	if edgePolicy == nil || !canBeHopHint {
+		return zpay32.HopHint{}, 0, false
+	}
+
+	if hopHintInfo.ScidAliasFeature {
+		alias, err := cfg.GetAlias(chanID)
+		_, included := alreadyIncluded[alias.ToUint64()]
+		switch {
+		case err != nil:
```

Comment:
my personal preference would be to have the error check for `GetAlias` directly after the call. Cause here you are accessing `alias.ToUint64()` before checking if the call even succeeded. This is fine at the moment since `GetAlias` does not return a pointer but if it was to change to a pointer then this would panic. 
Just a thought

Reply:
I like this a lot :+1: